### DISCUSSION TITLE
Any way to print the provided value when zod parse fails?

### DISCUSSION THREAD

**Discussion by jamischarles** [colinhacks/zod#1804] [Q&A] [answered]
Hi folks! I'm using Zod to validate API response shape in my node scripts.

Is there currently any built-in way to show a jest-style diff that shows expected value vs received value?

All I'm finding currently is ways to print the error message like "your input is too low".

I'm looking for:
`expected: number`
`provided: 'hello'`

Anything like this already exist in Zod?

Thanks!

**Comment by JacobWeisenburger** [comment]
I think what you want is [`.safeParse`](https://github.com/colinhacks/zod#safeparse)

```ts
const schema = z.string()
const result = schema.safeParse( 123 )
!result.success && console.log( result.error.issues )
// [
//     {
//       code: "invalid_type",
//       expected: "string",
//       received: "number",
//       path: [],
//       message: "Expected string, received number"
//     }
// ]
```

Let me know if you have any questions.

**Reply by jamischarles** [reply]
Thank you @JacobWeisenburger for the reply. 

Will this give me access to the input value?
```js
ie:
stringSchema.safeParse(12);
// => { success: false; error: ZodError }
// Q: Any way to get data in the error scenario ^?

stringSchema.safeParse("billie");
// => { success: true; data: 'billie' }
```

What I'm really after is the ability to see `data` when there's an error.

**Comment by JacobWeisenburger** [comment, marked as answer]
I don't think there is anything built into zod that does that. However, I have made this:
https://gist.github.com/JacobWeisenburger/d5dbb4d5bcbb287b7661061a78536423

**Comment by bakert** [comment]
Looks like this was added to zod for v4 https://zod.dev/error-customization?id=include-input-in-issues

```
z.string().parse(12, {
  reportInput: true
})
 
// ZodError: [
//   {
//     "expected": "string",
//     "code": "invalid_type",
//     "input": 12, // 👀
//     "path": [],
//     "message": "Invalid input: expected string, received number"
//   }
// ]
```

**Reply by wegry** [reply]
This should be marked as the answer. Works like a charm.

**Reply by samal-rasmussen** [reply]
Any way to do this globally? Having to add this to every single schema is tedious.

**Reply by bakert** [reply]
Seems like only monkey patching – https://github.com/colinhacks/zod/issues/5298#issuecomment-3353105776

**Reply by samal-rasmussen** [reply]
Thanks. I went with a patch to Zod using https://github.com/ds300/patch-package.

```
diff --git a/node_modules/zod/v4/core/util.cjs b/node_modules/zod/v4/core/util.cjs
index 8690887..f344e5b 100644
--- a/node_modules/zod/v4/core/util.cjs
+++ b/node_modules/zod/v4/core/util.cjs
@@ -569,9 +569,6 @@ function finalizeIssue(iss, ctx, config) {
     // delete (full as any).def;
     delete full.inst;
     delete full.continue;
-    if (!ctx?.reportInput) {
-        delete full.input;
-    }
     return full;
 }
 function getSizableOrigin(input) {
diff --git a/node_modules/zod/v4/core/util.js b/node_modules/zod/v4/core/util.js
index 1037513..6879ff4 100644
--- a/node_modules/zod/v4/core/util.js
+++ b/node_modules/zod/v4/core/util.js
@@ -512,9 +512,6 @@ export function finalizeIssue(iss, ctx, config) {
     // delete (full as any).def;
     delete full.inst;
     delete full.continue;
-    if (!ctx?.reportInput) {
-        delete full.input;
-    }
     return full;
 }
 export function getSizableOrigin(input) {
```
